Implementation for rmw_get_pub/sub_info_by_topic

dennis-adlink opened the pull request:
Implementation for the `rmw_get_publishers_info_by_topic` and `rmw_get_subscriptions_info_by_topic` methods, and some related refactoring:
- rewrite `rmw_collect_tptyp_for_kind` to the more generic `rmw_collect_data_for_endpoint`
- moved code from rmw_get_topic_endpoint_info.cpp to rmw_node.cpp (restructuring of rmw_node.cpp file is planned, but not in this PR)
- extracted qos conversion function `dds_qos_to_rmw_qos` from `get_readwrite_qos`

When this PR is merged, the rcl test `test_rcl_get_publishers_subscription_info_by_topic` can be enabled for Cyclone; I'll create a PR for this.

dennis-adlink commented:
This PR resolves issue #82 (may need refactoring in case the `Participant` mapping is moved to `Context`)

eboasson commented on rmw_cyclonedds_cpp/src/rmw_node.cpp:
```diff
@@ -2812,41 +2826,268 @@ extern "C" rmw_ret_t rmw_get_node_names(
+  rmw_topic_endpoint_info_array_t * endpoint_info)
+{
+  auto node_impl = static_cast<CddsNode *>(node->data);
+  const auto re_typ = std::regex("^(.*::)dds_::(.*)_$", std::regex::extended);
```
I think it would make sense to create a function to demangle topic names/types, with a parameter specifying whether it concerns data, requests or responses. There was some duplication already, but this adds yet another one and having regexes all over the place is a bit unpleasant.

dennis-adlink commented:
@eboasson Thanks for the review! I fixed the issues in my latest commit: 
- introduced a function for topic name/type demangle
- renamed `rd` to `ppant_rd`
- fixed cleaning up `topic_endpoint_info_array` in case of failure
- removed return value from `get_node_name` function
- refactored goto's in `get_readwrite_qos`
- removed `cont` from `rmw_collect_data_for_endpoint`

dennis-adlink commented:
One more commit to refactor out the extra call to `demangle_topic_type` in `get_endpoint_names_and_types_by_node`

eboasson commented:
Thanks @dennis-adlink. The only thing was a warning for an unused argument in a release build, I've pushed a fix for that.

@ivanpauno I think it can be merged, but perhaps you would like to have a quick look or do a CI run first?

ivanpauno commented:
> @ivanpauno I think it can be merged, but perhaps you would like to have a quick look or do a CI run first?

Yes, I will run CI together with the `rcl` PR, so we check that tests are passing.

ivanpauno commented on rmw_cyclonedds_cpp/src/rmw_node.cpp:
```diff
@@ -2812,41 +2826,268 @@ extern "C" rmw_ret_t rmw_get_node_names(
+      node_impl->enth, DDS_BUILTIN_TOPIC_DCPSPARTICIPANT, NULL, NULL)) < 0)
+  {
+    RMW_SET_ERROR_MSG("get_endpoint_info_by_topic: failed to create reader");
     return RMW_RET_ERROR;
```
If you first call [get_zero_initialized_topic_endpoint_info_array](https://github.com/ros2/rmw/blob/35ae2ff54317e8e6f30acace2fbaccd57fb2c566/rmw/src/topic_endpoint_info_array.c#L20), and then `init_with_size`, it shouldn't be a problem to call `rmw_topic_endpoint_info_array_fini` when something fails.

dennis-adlink commented:
> I think error handling should be fixed before merging, otherwise LGTM!

`rmw_topic_endpoint_info_array_init_with_size` does not set `count`, so that's why I initialize this to zero. After that point I use `rmw_topic_endpoint_info_array_fini` for cleaning up, also in case of an error (note that the code fragments above your remark is outdated). So I think error handling is properly in now.

ivanpauno commented:
> rmw_topic_endpoint_info_array_init_with_size does not set count, so that's why I initialize this to zero

I'm going to send a PR to fix that, because it's confusing (and incongruent with other arrays defined in `rcutils`/`rmw`). We can iterate on that in another PR, when the change in `rmw_topic_endpoint_info_array_init_with_size` is done.

> After that point I use rmw_topic_endpoint_info_array_fini for cleaning up, also in case of an error (note that the code fragments above your remark is outdated). So I think error handling is properly in now.

Sounds good!
